### DISCUSSION TITLE
Help with I2S to A2DP

### DISCUSSION THREAD

**Discussion by Kageni-wq** [pschatzmann/arduino-audio-tools#1410] [Q&A] [answered]
Hey, first thanks for this awesome resource.

I am lost in how to utilize different streams together and was hoping for some help. I have a ESP32 with I2S mic input, and I have a working sketch to transmit I2s input to A2DP. The targeted sounds I am attempting to auscultate are bettween 200 Hz to 2000 Hz, which aren't picked up well by a mems microphone. So I wanted to incorporate volume stream to allow boost to increase volume of I2S source as well as apply a low pass filter to filter frequencies >5kHz. I have tried incorporating a Volume Stream and a FilteredStream seperately at first, both of which aren't working as expected. 

First the volume stream, I get appropriate sounds through A2DP but altering volume value has no effect. 
#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

bool allow_boost = true;
AudioInfo info32(44100, 1, 32);
AudioInfo info16(44100, 2, 16);
BluetoothA2DPSource a2dp_source;
I2SStream i2s;
VolumeStream volume(i2s);
FormatConverterStream conv(&volume);
ConverterFillLeftAndRight<int32_t> bothChannels;
const int BYTES_PER_FRAME = 4;

int32_t get_sound_data(Frame* data, int32_t frameCount) {
  return conv.readBytes((uint8_t*)data, frameCount*BYTES_PER_FRAME)/BYTES_PER_FRAME;
}

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup conversion
  conv.begin(info32, info16);

  // start i2s input with default configuration
  Serial.println("starting I2S...");
  auto cfg = i2s.defaultConfig(RX_MODE);
  cfg.pin_bck = 27;
  cfg.pin_ws = 26;
  cfg.pin_data_rx = 15;
  cfg.i2s_format = I2S_PHILIPS_FORMAT; // or try with I2S_LSB_FORMAT  
  cfg.bits_per_sample = 32;
  cfg.sample_rate = 8000;
  cfg.channels = 1;
  cfg.buffer_count = 4;
  cfg.buffer_size = 8;
  cfg.copyFrom(info32);
  cfg.is_master = true;
  i2s.begin(cfg);

    // set initial volume
  volume.begin(cfg); // we need to provide the bits_per_sample and channels
  volume.setVolume(1.3);

  // start the bluetooth
  Serial.println("starting A2DP...");
  // a2dp_source.set_auto_reconnect(false);
  a2dp_source.start("", get_sound_data);

  Serial.println("A2DP started");

  a2dp_source.set_volume(255);
}

// Arduino loop - repeated processing
void loop() { delay(1000); }

Then the filtered stream, which does not produce any audio through A2DP, although it complies correctly and I can connect to bluetooth, I don't get any audio when connected. 

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

AudioInfo info32(44100, 1, 32);
AudioInfo info16(44100, 2, 16);
BluetoothA2DPSource a2dp_source;
I2SStream in;
I2SStream out;

FormatConverterStream conv(out);
ConverterFillLeftAndRight<int32_t> bothChannels;
const int BYTES_PER_FRAME = 4;

int32_t get_sound_data(Frame* data, int32_t frameCount) {
  return conv.readBytes((uint8_t*)data, frameCount*BYTES_PER_FRAME)/BYTES_PER_FRAME;
}

FilteredStream<int32_t, float> filtered(in, info32.channels);  // Assuming your filter has 64 coefficients
StreamCopy copier(out, filtered); // copies sound into i2s

// define FIR filter parameters
float coef[] = {
    0.000000000000000000,
    0.000001511297925605,
    0.000045825231132587,
    0.000101933246976758,
    0.000000000000000000,
    -0.000325252705726674,
    -0.000542934652017849,
    -0.000111225314686988,
    0.000973652950212339,
    0.001661972405759001,
    0.000587058394428470,
    -0.002140929286881939,
    -0.004003924383454057,
    -0.001928339829814701,
    0.003909210106703326,
    0.008340270816237191,
    0.005014346569892968,
    -0.006200715636442397,
    -0.015797621445604364,
    -0.011400390693660504,
    0.008736283732024654,
    0.028561496928001990,
    0.024508952141171887,
    -0.011075526033059554,
    -0.053639524213326052,
    -0.056754029493375331,
    0.012733664734546927,
    0.141263776411347297,
    0.267475614950405893,
    0.320009687542567145,
    0.267475614950405893,
    0.141263776411347297,
    0.012733664734546927,
    -0.056754029493375324,
    -0.053639524213326045,
    -0.011075526033059557,
    0.024508952141171898,
    0.028561496928001983,
    0.008736283732024655,
    -0.011400390693660513,
    -0.015797621445604371,
    -0.006200715636442400,
    0.005014346569892968,
    0.008340270816237200,
    0.003909210106703332,
    -0.001928339829814702,
    -0.004003924383454063,
    -0.002140929286881942,
    0.000587058394428470,
    0.001661972405759002,
    0.000973652950212338,
    -0.000111225314686988,
    -0.000542934652017848,
    -0.000325252705726674,
    0.000000000000000000,
    0.000101933246976758,
    0.000045825231132587,
    0.000001511297925605,
    0.000000000000000000,
};

// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup filters for all available channels
  filtered.setFilter(0, new FIR<float>(coef));
  filtered.setFilter(1, new FIR<float>(coef));

  // setup conversion
  conv.begin(info32, info16);

  // start i2s input with default configuration
  Serial.println("starting I2S...");
  auto cfg = in.defaultConfig(RX_MODE);  // Use 'in' for input stream
  cfg.pin_bck = 27;
  cfg.pin_ws = 26;
  cfg.pin_data_rx = 15;
  cfg.i2s_format = I2S_PHILIPS_FORMAT;
  cfg.bits_per_sample = 32;
  cfg.sample_rate = 44100;
  cfg.channels = 1;
  cfg.buffer_count = 4;
  cfg.buffer_size = 8;
  cfg.copyFrom(info32);
  cfg.is_master = true;
  in.begin(cfg);

  // start the bluetooth
  Serial.println("starting A2DP...");
  a2dp_source.start("", get_sound_data);

  Serial.println("A2DP started");

  a2dp_source.set_volume(255);
}

// Arduino loop - repeated processing
void loop() {
  copier.copy();
  }

**Comment by pschatzmann** [comment, marked as answer]
No, it's not working like this. You need to build a chain: e.g. I2SStream -> FormatConverterStream -> ConverterStream -> VolumeStream -> FilteredStream. In each constructor you indicate the source.
And the callback will then just read from the FilteredStream object.

No copy is needed/allowed in this chain.

The good thing is that you can build and test this step by step

**Reply by Kageni-wq** [reply]
Thanks for the reply, I tried updating the volume stream and I got it to compile but it is just producing loud static out of only the L side. Here is the updated sketch:

#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"

AudioInfo info32(16000, 1, 32);
AudioInfo info16(16000, 2, 16);
bool allow_boost = true;
BluetoothA2DPSource a2dp_source;
I2SStream i2s;
FormatConverterStream conv(i2s);
ConverterFillLeftAndRight<int32_t> bothChannels;
VolumeStream volume(conv);
const int BYTES_PER_FRAME = 4;


int32_t get_sound_data(Frame* data, int32_t frameCount) {
  return volume.readBytes((uint8_t*)data, frameCount*BYTES_PER_FRAME)/BYTES_PER_FRAME;
}



// Arduino Setup
void setup(void) {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // setup conversion
  conv.begin(info32, info16);

  // start i2s input with default configuration
  Serial.println("starting I2S...");
  auto cfg = i2s.defaultConfig(RX_MODE);
  cfg.pin_bck = 27;
  cfg.pin_ws = 26;
  cfg.pin_data_rx = 15;
  cfg.i2s_format = I2S_PHILIPS_FORMAT; // or try with I2S_LSB_FORMAT  
  cfg.bits_per_sample = 32;
  cfg.sample_rate = 16000;
  cfg.channels = 1;
  cfg.buffer_count = 4;
  cfg.buffer_size = 8;
  cfg.copyFrom(info32);
  cfg.is_master = true;
  i2s.begin(cfg);

  // Start Volume Stream
  volume.begin(cfg); // we need to provide the bits_per_sample and channels
  volume.setVolume(0.3);

  // start the bluetooth
  Serial.println("starting A2DP...");
  a2dp_source.set_auto_reconnect(true);
  a2dp_source.start("", get_sound_data);

  Serial.println("A2DP started");

  a2dp_source.set_volume(255);
}

// Arduino loop - repeated processing
void loop() { delay(1000); }

**Comment by pschatzmann** [comment]
Double check the bit sizes at each step: You define volume to be 32 bits, but after the conversion it's 16
Use the info16 and info32 that you have defined!
